### PR TITLE
user: test-wifi: associate by ssid

### PR DESCRIPTION
**Pull request by autrimpo**

### CODE REVIEW

File: user/uipcps/test-wifi.c [hunk 1]
```diff
@@ -640,6 +640,42 @@ wifi_add_network_to_config(struct wpa_ctrl *ctrl_conn,
...
+            return NULL;
+        }
+        if (!strncmp(ssid, ssid_conf, RL_WIFI_SSID_LEN)) {
+            id_len = strlen(id_conf);
+            id = malloc(id_len + 1);
+            if (!id) {
+                PE("Out of memory\n");
+                return NULL;
+            }
+            strncpy(id, id_conf, id_len + 1);
```
Comment (vmaffione) [inline user/uipcps/test-wifi.c]: You should use `strdup()` rather than strlen+malloc+strncpy

File: user/uipcps/test-wifi.c [hunk 2]
```diff
@@ -640,6 +640,42 @@ wifi_add_network_to_config(struct wpa_ctrl *ctrl_conn,
...
+    char id_conf[8];
+    char *id = NULL;
+    int id_len;
+
+    res = wpasup_send_cmd_get_resp(ctrl_conn, "LIST_NETWORKS");
+    p = res;
+
+    /* skip header */
+    while (*p != '\0' && *(p++) != '\n')
+        ;
```
Comment (vmaffione) [inline user/uipcps/test-wifi.c]: For code uniformity prefer `while (...) {}`

**Comment by vmaffione** [comment]
Thanks!

Documentation seems incomplete and inconsistent at fist sight:

- the usage of assoc is not in the documented (examples). In the end we foresee that this assoc command is the one that shim-wifi will really use, assuming an ID for the involved SSID already exists in the wpa_supplicant.conf.
- assoc calls ssid_to_id(), then NET_DISABLE, NET_ENABLE ID, RECONNECT and WAIT FOR CONNECTED
- Examples say that in order to connect to a WPA2 network you need to first to ADD_NETWORK SSID, then LIST_NETWORK (to see what is the ID corresponding to the SSID), and then call NET_ENABLE ID. So RECONNECT and WAIT FOR CONNECTED are not called in this case. Why this difference?

**Comment by autrimpo** [comment]
Yes, the documentation is inconsistent.

The documentation (and code) I wrote as a part of previous PR assumed interactive use of the tool by the end user to manage the network configurations. The *connect to WPA2* example is a minimal example starting from clean slate that achieves the desired result. Enabling the first network configuration means wpa_supplicant will try to associate with it. `RECONNECT` is therefore not needed.

The problem is, `RECONNECT` does not take any arguments, so if we have multiple configurations enabled, we cannot be sure which network we actually connect to. That's the reason for the `NET_DISABLE all`, `NET_ENABLE ID` at the beggining of the connection - to ensure determinism. `RECONNECT` isn't actually needed here either, as it's the `NET_ENABLE` that goes from 0 to 1 enabled networks that initiates the connection, it's there just to make "double sure" we're associating with a network.

We can't be sure that enabling a network (in general) will initiate association either - maybe we're already connected to a network, so there's no `wait_for_connected` in `NET_ENABLE`. And since there's no separate command for it, the example doesn't call it.

It was my understanding that the examples at the top are for people (for example, you) who would want to try the tool, not documentation of the shim-wifi interface. I can add a section that documents the three interface commands that you've mentioned in the last PR in greater detail.

**Comment by vmaffione** [comment]
Thanks for the explanation, that makes sense!
The documentation for tool is nice and can stay there, but what we actually need is the shim-wifi to have a way to associate a given SSID.

So I guess what we need is just a second ittle section in the documentation/examples about how the shim-wifi should use the library you are writing.
I guess that would be:
 - a minimal example of wpa_supplicant that includes a "configuration entry" for an SSID with PDK.
 - an example of assoc command that connects to that SSID

**Pull request merged**